File: maratona-8-superior/d_aritmetica/d_aritmetica-3.cpp
#include <bits/stdc++.h>

using namespace std;

int main () {


    string operacao;
    vector<string> nomes;
    vector<int> numeros;
    nomes.push_back("um");
    nomes.push_back("dois");
    nomes.push_back("tres");
    nomes.push_back("quatro");
    nomes.push_back("cinco");
    nomes.push_back("seis");
    nomes.push_back("sete");
    nomes.push_back("oito");
    nomes.push_back("nove");
    nomes.push_back("dez");
    nomes.push_back("onze");
    nomes.push_back("doze");
    nomes.push_back("treze");
    nomes.push_back("catorze");
    nomes.push_back("quatorze");
    nomes.push_back("quinze");
    nomes.push_back("dezesseis");
    nomes.push_back("dezessete");
    nomes.push_back("dezoito");
    nomes.push_back("dezenove");
    nomes.push_back("vinte");
    for(int i=1; i<=14; i++)
        numeros.push_back(i);
    numeros.push_back(14);
    for(int i=15; i<=20; i++)
        numeros.push_back(i);

        vector<string> lista;
    vector<int> soma;
    while(cin >> operacao){
        //cout << operacao << "\n";
        //cout << lista.size() << "\n";
        lista.push_back(operacao);
    }
    if(lista.size() == 1)
        cout << lista[0] << "\n";
    else{
        for(int i = 1; i < lista.size(); i+=2){
            if(lista[i] == "vezes"){
                int acum ;
                if(i > 2){
                    if(lista[i-2] == "mais" )
                        acum = 1;
                    else if(lista[i-2] == "menos" )
                        acum = -1;


                }
                else
                    acum = 1;
                acum *= (numeros[find(nomes.begin(), nomes.end(), lista[i-1])-nomes.begin()]);
                if(i > 2)
                    lista[i-2] = "";
                //cout << acum << "|";
                for(; i < lista.size() && lista[i] == "vezes"; i+=2){
                    acum *= (numeros[find(nomes.begin(), nomes.end(), lista[i+1])-nomes.begin()]);
                    //cout << acum << "|";
                    lista[i+1] = "";
                    lista[i] = "";
                    lista[i-1] = "";
                }

                soma.push_back(acum);
            }
        }
        //cout << "passou\n";
        if(lista[0] != "")
            soma.push_back(numeros[find(nomes.begin(), nomes.end(),lista[0]) - nomes.begin()]);
        for(int i = 1; i < lista.size(); i++){

            //cout << "&\n";
            if(lista[i] == "mais"){
                //soma.push_back(numeros[find(nomes.begin(), nomes.end(),lista[i-1]) - nomes.begin()]);
                soma.push_back(numeros[find(nomes.begin(), nomes.end(),lista[i+1]) - nomes.begin()]);
                i++;
            }
            if(lista[i] == "menos"){
                //soma.push_back((-1)*(numeros[find(nomes.begin(), nomes.end(),lista[i-1]) - nomes.begin()]));
                soma.push_back((-1)*(numeros[find(nomes.begin(), nomes.end(),lista[i+1]) - nomes.begin()]));
                i++;
            }
        }

        int acum = 0;
        for(int i = 0 ; i < soma.size(); i++)
            acum += soma[i];

        cout << nomes[find(numeros.begin(), numeros.end(),acum)-numeros.begin()] << "\n";
    }



    return 0;
}

File: 08-maratona/superior/d_aritmetica/d_aritmetica-3.py
my_dict = {"um": 1, "dois": 2, "três": 3, "quatro": 4, "cinco": 5, 
           "seis": 6, "sete": 7, "oito": 8, "nove": 9, "dez": 10,
           "onze": 11, "doze": 12, "treze": 13, "catorze": 14, "quatorze": 14, "quinze": 15,
           "dezesseis": 16, "dezessete": 17, "dezoito": 18, "dezenove": 19, "vinte": 20}
my_dict_reversed = {k: j for j, k in my_dict.items()}

entry = input().split()
calc = [my_dict[j] if i%2 == 0 else j for i, j in enumerate(entry)]
my_operand, my_operators, skip = [], [], False
#Vezes
for i in range(len(calc)):
  if skip:
    skip = False
    continue
  if calc[i] not in ("mais", "menos", "vezes"):
    my_operand.append(calc[i])
  else:
    if calc[i] == "vezes":
      my_operand.append(my_operand.pop(-1) * calc[i+1])
      skip = True
    else:
      my_operators.append(calc[i])
#Mais / Menos
for i in my_operators:
  if i == "mais":
    my_operand.append(my_operand.pop() + my_operand.pop())
  else:
    my_operand.append(abs(my_operand.pop() - my_operand.pop()))
print(my_dict_reversed[my_operand[0]])